tensorflow/core/ops/math_ops.cc

```diff
@@ -1447,4 +1447,56 @@ b: Another tensor, of same type and shape as `a`.
 product: Pairwise cross product of the vectors in `a` and `b`.
 )doc");
 
+// --------------------------------------------------------------------------
+
+REGISTER_OP("Cumsum")
+    .Input("x: T")
+    .Input("axis: int32")
+    .Attr("reverse: bool")
+    .Output("out: T")
+    .Attr("T: numbertype")
+    .Doc(R"doc(
+Compute the cumulative sum of the tensor `x` along `axis`.
+
+The output `out` at any given index i is equal to the sum of all
+elements `x_j` of `x` with j <= i.
+
+By setting the `reverse` operation to `True`, the sum is performed in the
+reverse order. In contrast to using `tf.reverse`, this avoids copying the
+tensor.
+
+For example:
+
+```prettyprint
+# tensor 'x' is [1, 2, 3, 4, 5]
+tf.cumsum(x)               ==> [1, 3, 6, 10, 15]
+tf.cumsum(x, reverse=True) ==> [5, 9, 12, 14, 15]
```

Comment:
@girving I believe the `reverse=True` output is correct. It cumulatively adds up the tensor starting from the end and iterates backwards, but it doesn't reverse the output tensor:

```
cumsum([1, 2, 3, 4, 5])
= [(1), (1 + 2), (1 + 2 + 3), (1 + 2 + 3 + 4), (1 + 2 + 3 + 4 + 5)]
= [1, 3, 6, 10, 15]

cumsum([1, 2, 3, 4, 5], reverse=True)
= [(5), (5 + 4), (5 + 4 + 3), (5 + 4 + 3 + 2), (5 + 4 + 3 + 2 +1)] 
= [5, 9, 12, 14, 15]
```